File: 2021_lua/day11.lua
-- Advent Of Code 2021
-- Copyright (C) 2024  Lukas Heindl
-- 
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
-- 
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
-- 
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

local colors = require("term.colors")
local term   = require("term")

local _M = {}

local function parse_file(file)
	local field = {}
	for line in file:lines() do
		local row = {}
		for x in line:gmatch(".") do
			table.insert(row, tonumber(x))
		end
		table.insert(field, row)
	end
	return field
end

local function print_field(field, pre)
	if graphic then
		term.clear()
		term.cursor['goto'](1, 1)
		if pre then print(pre) end
	end
	for _,r in ipairs(field) do
		for _,c in ipairs(r) do
			if c == 0 then
				io.write(colors.red(string.format("%1d", c)))
			else
				io.write(string.format("%1d", c))
			end
		end
		io.write("\n")
	end
end

local function step(field)
	local amount = 0
	local function inc(y,x)
		if not field[y] or not field[y][x] then return end

		field[y][x] = field[y][x] + 1

		if field[y][x] == 10 then
			amount = amount + 1
			inc(y+0,x+1)
			inc(y+0,x-1)
			inc(y+1,x+0)
			inc(y+1,x+1)
			inc(y+1,x-1)
			inc(y-1,x+0)
			inc(y-1,x+1)
			inc(y-1,x-1)
		end
	end
	for y,_ in ipairs(field) do
		for x,_ in ipairs(field[y]) do
			inc(y,x)
		end
	end
	for y,_ in ipairs(field) do
		for x,_ in ipairs(field[y]) do
			if field[y][x] > 9 then field[y][x] = 0 end
		end
	end
	return amount
end

function _M.part1(file)
	local field = parse_file(file)
	if graphic then print_field(field, "Part1:") end

	local sum = 0
	for i=1,100 do
		sum = sum + step(field)
		if graphic then
			print_field(field, "Part2:")
			os.execute("sleep 0.05")
		end
	end
	return sum
end

function _M.part2(file)
	local function all_flash(field)
		for _,r in ipairs(field) do
			for _,c in ipairs(r) do
				if c ~= 0 then return false end
			end
		end
		return true
	end
	local field = parse_file(file)
	if graphic then print_field(field, "Part2:") end

	local i = 1
	while true do
		step(field)
		if graphic then
			print_field(field, "Part2:")
			os.execute("sleep 0.01")
		end
		if all_flash(field) then return i end
		i = i + 1
	end
	return nil
end

return _M

File: 2020_C/day-11.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "./read_file.h"

/*
Advent Of Code 2020
Copyright (C) 2024  Lukas Heindl

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// returns -1 on out of bounds
// returns  0 on free seat
// returns  1 on occupied seat
int check_seat1(char** arr, int length, int width, int x, int y, int step_x, int step_y){
	if(x + step_x >= 0 && x + step_x < width  && y + step_y >= 0 && y + step_y < length){
		if(arr[y+step_y][x+step_x] == '#'){
			/* printf("0\n"); */
			return 1;
		} else {
			/* printf("1\n"); */
			return 0;
		}
	} else {
		// out of bounds
		/* printf("-1\n"); */
		return -1;
	}
}
int check_seat2(char** arr, int length, int width, int x, int y, int step_x, int step_y){
	while(x + step_x >= 0 && x + step_x < width  && y + step_y >= 0 && y + step_y < length){
		if(arr[y+step_y][x+step_x] == '#'){
			return 1;
		}
		if(arr[y+step_y][x+step_x] == 'L'){
			return 0;
		}
		x += step_x;
		y += step_y;
	}
	return 0;
}

int part(struct string_content* input1, struct string_content* input2, int tolerance, int check_seat(char**, int, int, int, int, int, int)){
	int width = strlen(input1->content[0]);
	/* print_file(input1); */

	// recognize if map changed
	int changed = 1;

	
	while(changed){
		changed = 0;
		// cycle over the lines and chars in lines
		for(int i=0; i < input1->length; i++){
			for(int j=0; j < width; j++){
				// check if someone will chose this seat in this iteration
				if (	(input1->content[i][j] == 'L') && // seat is empty
						/* // linear */
						check_seat(input1->content, input1->length, width, j, i, +0, +1) <= 0 &&
						check_seat(input1->content, input1->length, width, j, i, +0, -1) <= 0 &&
						check_seat(input1->content, input1->length, width, j, i, +1, +0) <= 0 &&
						check_seat(input1->content, input1->length, width, j, i, -1, +0) <= 0 &&
						/* // diagonal */
						check_seat(input1->content, input1->length, width, j, i, +1, -1) <= 0 &&
						check_seat(input1->content, input1->length, width, j, i, -1, -1) <= 0 &&
						check_seat(input1->content, input1->length, width, j, i, +1, +1) <= 0 &&
						check_seat(input1->content, input1->length, width, j, i, -1, +1) <= 0){
					/* printf("Seat %d,%d is surrounded by not occupied seats.\n", i, j); */
					input2->content[i][j] = '#'; // occupy seat
					changed = 1;
				}
				// check if this seat will be left in this iteration
				else if (input1->content[i][j] == '#'){
					int amount = 0;
					// linear
					if(check_seat(input1->content, input1->length, width, j, i, +0, +1) == 1){ /*if(i==2 && j == 7) printf("01\n");*/ amount++; }
					if(check_seat(input1->content, input1->length, width, j, i, +0, -1) == 1){ /*if(i==2 && j == 7) printf("02\n");*/ amount++; }
					if(check_seat(input1->content, input1->length, width, j, i, +1, +0) == 1){ /*if(i==2 && j == 7) printf("03\n");*/ amount++; }
					if(check_seat(input1->content, input1->length, width, j, i, -1, +0) == 1){ /*if(i==2 && j == 7) printf("04\n");*/ amount++; }
					// diagonal
					if(check_seat(input1->content, input1->length, width, j, i, +1, -1) == 1){ /*if(i==2 && j == 7) printf("05\n");*/ amount++; }
					if(check_seat(input1->content, input1->length, width, j, i, -1, -1) == 1){ /*if(i==2 && j == 7) printf("06\n");*/ amount++; }
					if(check_seat(input1->content, input1->length, width, j, i, +1, +1) == 1){ /*if(i==2 && j == 7) printf("07\n");*/ amount++; }
					if(check_seat(input1->content, input1->length, width, j, i, -1, +1) == 1){ /*if(i==2 && j == 7) printf("08\n");*/ amount++; }

					if(amount >= tolerance){
						input2->content[i][j] = 'L';
						changed = 1;
					}
				} // close else if
			} // close for loop j
		} // close for loop i

		// push update to input1
		for(int i=0; i < input1->length; i++){
			strcpy(input1->content[i], input2->content[i]);
		}
		/* printf("\n"); */
		/* print_file(input1); */

	} // end of while

	// count occupied places
	int counter = 0;
	for(int i=0; i < input1->length; i++){
		for(int j=0; j < width; j++){
			if(input1->content[i][j] == '#'){
				counter++;
			}
		}
	}
	return counter;
}

int main() {
	char* type;
	(void)type;

	struct string_content *input1;
	struct string_content *input2;
	input1 = read_file("./problems/day-11.dat", type);
	input2 = read_file("./problems/day-11.dat", type);
	printf("Part1:\nAmount of occupied seats (Prediction 1): %d\n", part(input1, input2, 4, check_seat1));
	free_file(input1);
	free_file(input2);
	
	input1 = read_file("./problems/day-11.dat", type);
	input2 = read_file("./problems/day-11.dat", type);
	printf("\nPart2:\nAmount of occupied seats (Prediction 2): %d\n", part(input1,input2, 5, check_seat2));
	free_file(input1);
	free_file(input2);
	return EXIT_SUCCESS;
}
